Shion recurring session

nowknowing opened the pull request:
1. In SessionDate: added hasSameTime and calculateDaysBetween (gives calendar not 24-hr days) static methods.
2. RecurringSession class added.
- hasSessionOn is meant for sheduler to check if session happening on given day. lastSessionOnOrBefore will then retrive the particular Session. [@yungweezy  ChoonWei]

- numOfSessionBetween counts the number of session btw. Meant for fee calculation. [@enhao25 Enhao]

3. Light Test. edit dates to check.
Primarily tests the logic-heavy numOfSessionBetween method.

TODO: non-temporary tests.

nowknowing committed:
Recurring session logic still need fixing.

nowknowing committed:
1. In SessionDate: added hasSameTime and calculateDaysBetween (gives calendar not 24-hr days) static methods.
2. RecurringSession class added.
hasSessionOn is meant for sheduler to check if session happening on given day. lastSessionOnOrBefore will then retrive the particular Session.
numOfSessionBetween counts the number of session btw. Meant for fee calculation.
3. Light Test that you can edit as you like the input dates.
Primarily tests the logic-heavy numOfSessionBetween method.

TODO: submittable tests.

nowknowing committed:
Merge conflict. accidently deleted 2 getters, so retrived them

nowknowing committed:
Merge branch 'master' into shion_recurring_session

enhao25 commented on build.gradle:
```diff
@@ -45,6 +45,7 @@ task coverage(type: JacocoReport) {
 }
 
 dependencies {
+    implementation 'org.jetbrains:annotations:20.1.0'
```
Yo actually what is this for ah?

nowknowing replied:
no idea myself leemme delete

enhao25 commented on src/test/java/seedu/address/model/session/RecurringSessionTest.java:
```diff
@@ -0,0 +1,31 @@
+        SessionDate firstOfMarch = new SessionDate("2021-03-01", "00:00");
+        SessionDate lastOfMarch = new SessionDate("2021-03-31", "23:59");
+        int ans = s.numOfSessionBetween(firstOfMarch, lastOfMarch);
+        System.out.println(ans);
```
Hi, understand the the testing is still a work in progress, but I think it could be better if we didn't use System.out.println() when we are merging to master. Using assertions might be better.

nowknowing replied:
right!

enhao25 commented on src/main/java/seedu/address/model/session/Interval.java:
```diff
@@ -0,0 +1,46 @@
+import static seedu.address.commons.util.AppUtil.checkArgument;
+
+/**
+ * Represents a Recurring Session's interval.
```
LGTM. But was thinking if we can be more explicit that the interval is in terms of the number of days.

nowknowing replied:
yep!

enhao25 commented on src/main/java/seedu/address/model/session/RecurringSession.java:
```diff
@@ -0,0 +1,130 @@
+
+    /**
+     * Constructs a {@code RecurringSession}.
+     * <p>
```
Hi was just wondering what could does <p> mean. Isit a typo?

enhao25 commented on src/main/java/seedu/address/model/session/RecurringSession.java:
```diff
@@ -0,0 +1,130 @@
+        requireAllNonNull(interval, endDateTime);
+        this.interval = interval;
+        checkArgument(isValidEnd(sessionDate, endDateTime, interval), MESSAGE_CONSTRAINTS);
+        this.endDateTime = endDateTime;
```
Hi, just checking, for this implementation, we run super(session) to create a standalone session out, but then it doesn't seem to keep a reference to the created session anymore. Can I check if that is the intended effect?

nowknowing replied:
No it's not intended because I didn't (and kind of currently dont) know what problems this may cause.
Not intended as in never questioned it
What problems might this cause/ what recommended way might there be?
Was thinking since its extending Session, the parent's method to access Sesssion contents are inherited

enhao25 replied:
Hmmmm already, actually didn't use super often so not sure how it would work out. We could keep this in mind and when the rest of us work on features that uses recurring session, see if we are able to perform all the actions we want, even w/o keeping a reference to it ba.

enhao25 reviewed (commented):
Hi, good job! Left some comments regarding the code. Also I am just wondering if you could include a way (Probably through a command) to add a recurring session into to application, so that the reviewers can checkout and run the PR to perform manual testing as well, to better understand your implementation. Thanks!

nowknowing committed:
1. remove annotation dependency of build.gradle
2. Improve description of Interval
3. Add assertion to onSessionDate and requireNonNulls to all methods
4. use SessionDate existing methods
5. 2 tests using assertEquals() though expected is hard-coded(eye-balled)

nowknowing committed:
Merge branch 'master' of https://github.com/AY2021S2-CS2103T-T11-1/tp into shion_recurring_session

# Conflicts:
#	src/main/java/seedu/address/model/session/SessionDate.java

nowknowing committed:
Merge branch 'shion_recurring_session' of https://github.com/nowknowing/tp into shion_recurring_session

# Conflicts:
#	src/main/java/seedu/address/model/session/SessionDate.java

nowknowing committed:
fix merge conflict lost items. fix recurringsession constructor typo

nowknowing committed:
fix import order

enhao25 commented on src/main/java/seedu/address/model/session/SessionDate.java:
```diff
@@ -60,6 +61,20 @@ public LocalTime getTime() {
     }
 
+    public static boolean hasSameTime(SessionDate s1, SessionDate s2) {
+        return s1.getTime().equals(s2.getTime());
```
Actually not sure if other groups will nitpick this, but we could give s1 and s2 a more meaningful name such as sessionDate1 and sessionDate2. Will be longer, but at least we prevent other group from nitpicking these small bugs ah.

nowknowing replied:
Yep thanks. actualy given your comments I think there was also no need for this to be a static method.
Will replace with a .equalTime(SessionDate sessionDate)

enhao25 commented on src/test/java/seedu/address/model/session/RecurringSessionTest.java:
```diff
@@ -0,0 +1,40 @@
+
+    @Test
+    void getEndDateTime() {
+    }
```
Hi, maybe we could add a //TODO: tag here, so that we would remember to come back here in the near future and work on this.

nowknowing replied:
Thanks, did in the coming commit

enhao25 commented on src/main/java/seedu/address/model/session/RecurringSession.java:
```diff
@@ -0,0 +1,140 @@
+     * @param interval
+     * @param endDateTime must match sessionDate and interval.
+     */
+    public RecurringSession(SessionDate sessionDate, Duration duration, Subject subject, Fee fee,
```
Also I am just wondering if there could be a way (Probably through a command) to add a recurring session into to application, so that the reviewers can checkout and run the PR to perform manual testing as well, to better understand your implementation. Not sure if you want to work on that first before we reveal and approve this to ensure that the implementation passes some light testing. If not it might be hard for me and jonah to test if our feature works if we couldn't add recurring session inside to the addressbook to test.

nowknowing replied:
yep will do this

enhao25 reviewed (commented):
Hi! Nice work on making the edits, had some additional comments that could be open for discussion.

nowknowing committed:
Merge branch 'master' of https://github.com/AY2021S2-CS2103T-T11-1/tp into shion_recurring_session

nowknowing committed:
edit SessionDate methods to non-static. Reorder RecurringSession methods

nowknowing committed:
1. RecurringSession extends Session. So does RecurringSessionCommand. but NOT RecurringSessionCommandParser.
2. edit: AddSessionCommand equal method wasn't checking for name

TODO: edit TuitionCard to reflect "Recurring" tag or something.

nowknowing committed:
Merge branch 'master' of https://github.com/AY2021S2-CS2103T-T11-1/tp into shion_recurring_session

nowknowing committed:
Add comment "TODO"

nowknowing committed:
Add todo comment

nowknowing committed:
Add Todo comment for UI in tuition card

nowknowing committed:
Add comments and fix minor problems

nowknowing committed:
Rename all endDate to LastDate to prevent clash with Choonwei's overlap checks.
Edit Delete Session to reflect recurring session was deleted.

nowknowing committed:
Edit comment mismatch

enhao25 commented on src/main/java/seedu/address/logic/commands/AddRecurringSessionCommand.java:
```diff
@@ -0,0 +1,70 @@
+            + PREFIX_TIME + "12:00 "
+            + PREFIX_DURATION + "90 "
+            + PREFIX_SUBJECT + "Math "
+            + PREFIX_FEE + "40"
```
LGTM, I think "40" is missing a space, to be "40 "

enhao25 commented on src/main/java/seedu/address/logic/commands/AddRecurringSessionCommand.java:
```diff
@@ -0,0 +1,70 @@
+    /**
+     * Creates an AddRecurringSessionCommand to add the specified {@code RecurringSession}
+     */
+    public AddRecurringSessionCommand(RecurringSession recurringSession, Name name) {
```
Maybe you would want to pull upstream first and merge inside, I believe that choon wei added a functionally that previously adding session that clashes. (E.g. If a session that starts on 2020-01-01 12.00 that last for 120mins, we can't create another session that is between  2020-01-01 12.00 to 2020-01-01 13.59) Although, with recurring session, this is much harder to achieve, because we need to take into consideration from the start to end date, if adding this rec_session will clash with future individual sessions as well. Not sure if you want to push the 2nd part into a separate issue as it seems like quite a big implementation.

nowknowing replied:
Yeah ive pulled upstream. ChoonWei's method doesn't check for Recurring session so there arent clashes.
Yep will do as a seperate issue

nowknowing committed:
Merge branch 'master' of https://github.com/AY2021S2-CS2103T-T11-1/tp into shion_recurring_session
TODO : Overlap check for recurring session

enhao25 commented on src/main/java/seedu/address/model/session/RecurringSession.java:
```diff
@@ -0,0 +1,167 @@
+/**
+ * Class that handles RecurringSession that extend Session.
+ */
+public class RecurringSession extends Session {
```
Ohya one of the possible issue is that I think the storage has no way of identifying now whether a session is a recurring session or a standalone session. I tried creating the recurring session, close the app and open it again, and the recurring session becomes a standalone session now. Might need to take the storage into consideration, either in this iteration, or as a separate issue.

nowknowing replied:
Right ive got to make a RecurringSession json. to save things into if Session instanceOf RecurringSession.

nowknowing replied:
Just done so, I think should be working

enhao25 reviewed (commented):
Hi! Thanks for working so hard for this functionality. Have added some comments after doing some light manual testing. Some of the issue might be quite heavy, so do consider if you want to work on it during this iteration or as a separate issue.

nowknowing committed:
Merge branch 'master' of https://github.com/AY2021S2-CS2103T-T11-1/tp into shion_recurring_session

nowknowing committed:
Added JsonAdaptedRecurringSession extending JsonAdaptedSession

nowknowing committed:
Fix Checkstyle errors

nowknowing committed:
add space to "40 " i.e. debug

nowknowing committed:
Forgot to add this to the commit.

nowknowing committed:
Merge branch 'master' of https://github.com/AY2021S2-CS2103T-T11-1/tp into shion_recurring_session

# Conflicts:
#	src/main/java/seedu/address/logic/parser/AddressBookParser.java

enhao25 commented on src/main/java/seedu/address/model/session/RecurringSession.java:
```diff
@@ -0,0 +1,168 @@
+ */
+public class RecurringSession extends Session {
+    public static final String MESSAGE_CONSTRAINTS = "Last date/time does not match starting date and interval.";
+    private Interval interval;
```
Hi thanks for the hard work. One of the bugs I found was that if I were to put a value where the end date and start date doesn't match, instead of showing an error message, an exception occurs. I think for this, it would be better to show some form of error message to the user instead.
![image](https://user-images.githubusercontent.com/24240408/112743666-df27e980-8fcb-11eb-9d65-1d072aa0ea22.png)

![image](https://user-images.githubusercontent.com/24240408/112743685-04b4f300-8fcc-11eb-8931-237e9f17ec13.png)

nowknowing replied:
Thanks have added this as a new separate issue

enhao25 commented on src/main/java/seedu/address/storage/JsonAdaptedRecurringSession.java:
```diff
@@ -0,0 +1,58 @@
+     * @param lastSessionDate
+     */
+    @JsonCreator
+    public JsonAdaptedRecurringSession(@JsonProperty("sessionDate") String sessionDate,
```
Hi, so sorry, but I am not sure where the error is.

However, if you run the command:
add_rec_session n/Alex Yeoh d/2020-03-03 t/12:00 k/90 s/Math f/40 b/7 e/2020-03-17

then close the application, open the application again.

Run the command:
delete_session n/Alex Yeoh i/<number of the recurring session>
You will notice that the output will be "Deleted Session: Date: 2020-03-03; Time: 12:00; Duration: 90; Subject: Math; Fee: 40.0" instead of "Deleted Recurring Session: Date: 2020-03-03; Time: 12:00; Duration: 90; Subject: Math; Fee: 40.0; Interval: 7; Last Date: 2020-03-17".

Hence, the although the storage is storing the interval and last session value inside the json file, I think when the system reads from the storage, it creates a session instead of a recurring session. Might need to look into that.

nowknowing replied:
Ok found it will fix

nowknowing replied:
actually no idea

nowknowing replied:
Issue is solved by adding type information in json, as necessary for deserialisation polymorphism.

enhao25 reviewed (commented):
Hi! Good job. Left some comments on some of the bugs found for the current implementation.

nowknowing committed:
Solved deserialisation polymorphism issue by adding type info in JsonAdapted -Sessions/RecurringSessions.

(Not with pre-filled addressbook since the session included does not reflect this change. Use fresh new empty addressbook)

nowknowing committed:
fix checkstyle

nowknowing committed:
Edit JsonSerializableAddressBookTest data to include type for session

nowknowing committed:
delete custom type name for consistency

enhao25 reviewed (approved):
Yeap LGTM.

samleewy commented on src/main/java/seedu/address/model/session/RecurringSession.java:
```diff
@@ -0,0 +1,168 @@
+    // Checks that lastDate can be on given sessionDate.
+    private static boolean isValidEnd(SessionDate firstSessionDate, SessionDate lastSessionDate, Interval interval) {
+        boolean occursAtTime = firstSessionDate.isSameTime(lastSessionDate);
+        boolean occursAtDate = occursAtDate(firstSessionDate, lastSessionDate, interval);
```
probably could rename it to `boolean hasOccuredAtDate`, since the name seems to clash with the method name.

nowknowing replied:
ok, in coming commit!

samleewy commented on src/main/java/seedu/address/model/session/RecurringSession.java:
```diff
@@ -0,0 +1,168 @@
+
+    // Checks that lastDate can be on given sessionDate.
+    private static boolean isValidEnd(SessionDate firstSessionDate, SessionDate lastSessionDate, Interval interval) {
+        boolean occursAtTime = firstSessionDate.isSameTime(lastSessionDate);
```
same for this as well, suggest to rename to `hasOccuredAtTime`

samleewy commented on src/main/java/seedu/address/storage/JsonAdaptedSession.java:
```diff
@@ -10,6 +12,12 @@
 import seedu.address.model.session.SessionDate;
 import seedu.address.model.session.Subject;
 
+@JsonTypeInfo(
+        use = JsonTypeInfo.Id.NAME,
+        include = JsonTypeInfo.As.PROPERTY,
+        property = "type")
+@JsonSubTypes(
+        @JsonSubTypes.Type(value = JsonAdaptedRecurringSession.class))
```
Thanks for figuring this out! neat!!

samleewy commented on src/main/java/seedu/address/model/session/RecurringSession.java:
```diff
@@ -0,0 +1,168 @@
+            firstInSpan = new SessionDate(firstInSpanLdt.toString());
+        }
+
+        return firstInSpan.numOfDayTo(lastInSpan) / interval.getValue() + 1;
```
I'm assuming it's `(firstInSpan.numOfDayTo(lastInSpan) / interval.getValue()) + 1`. might be better to put brackets!

samleewy reviewed (approved):
Thanks for the hardwork!!
Some minor changes that's required, and also the pending #132.
Feel free to merge once you're done!

chwoozy reviewed (approved):
Other than the current comments, LGTM! 👍🏼 
Pulled and tried on local and works fine too.

nowknowing committed:
Renamed OccuredAt to more boolean-obvious and informative method name.
Add bracket for clarity

nowknowing committed:
checkstyle fix
